File: schemas/typescript/validation.ts
import { sharedSchema } from "../generated/typescript/schema.js";

export interface ValidationOptions {
  knownExtensions?: readonly string[];
}

export interface SchemaValidationError {
  code: string;
  path: string;
}

export type ValidationResult<T> =
  | {
      ok: true;
      value: T;
    }
  | {
      ok: false;
      error: SchemaValidationError;
    };

export class SchemaValidationException extends Error {
  readonly error: SchemaValidationError;

  constructor(error: SchemaValidationError) {
    super(`${error.code} at ${error.path}`);
    this.name = "SchemaValidationException";
    this.error = error;
  }
}

type SchemaNode = Record<string, unknown>;
type JsonRecord = Record<string, unknown>;

const schema = sharedSchema as unknown as SchemaNode;

export function validateDefinition(
  definition: string,
  input: unknown,
  options: ValidationOptions = {},
): ValidationResult<unknown> {
  const definitions = asRecord(schema.$defs);
  const definitionSchema = definitions?.[definition];
  if (!isSchemaNode(definitionSchema)) {
    return failure("schema-definition-unavailable", "$");
  }

  const result = validateNode(definitionSchema, input, "$", schema);
  if (!result.ok) {
    return result;
  }

  if (definition === "PluginManifest") {
    return validateRequiredExtensions(input, options.knownExtensions ?? []);
  }

  return success(input);
}

export function parseWithDefinition<T>(
  definition: string,
  input: unknown,
  options: ValidationOptions = {},
): T {
  const result = validateDefinition(definition, input, options);
  if (!result.ok) {
    throw new SchemaValidationException(result.error);
  }
  return result.value as T;
}

export function migratePluginManifest(
  input: unknown,
  options: ValidationOptions = {},
): { migrated: boolean; value: unknown } {
  if (!isRecord(input)) {
    throw new SchemaValidationException({
      code: "schema-type-mismatch",
      path: "$",
    });
  }
  if (typeof input.schemaVersion !== "string") {
    throw new SchemaValidationException({
      code: "schema-required-field",
      path: "$.schemaVersion",
    });
  }

  const migrations = Array.isArray(schema["x-openab-migrations"])
    ? schema["x-openab-migrations"]
    : [];
  const migration = migrations.find(
    (candidate) =>
      isRecord(candidate) &&
      candidate.definition === "PluginManifest" &&
      candidate.fromSchemaVersion === input.schemaVersion,
  );

  if (isRecord(migration)) {
    const value = cloneJson(input);
    const rename = asRecord(migration.rename) ?? {};
    for (const [from, target] of Object.entries(rename)) {
      if (typeof target !== "string" || !(from in value)) {
        continue;
      }
      if (target in value) {
        throw new SchemaValidationException({
          code: "schema-migration-conflict",
          path: "$",
        });
      }
      value[target] = value[from];
      delete value[from];
    }
    value.schemaVersion = migration.toSchemaVersion;

    const result = validateDefinition("PluginManifest", value, options);
    if (!result.ok) {
      throw new SchemaValidationException(result.error);
    }
    return { migrated: true, value };
  }

  if (input.schemaVersion !== schema["x-openab-schema-version"]) {
    throw new SchemaValidationException({
      code: "schema-migration-unavailable",
      path: "$.schemaVersion",
    });
  }

  const result = validateDefinition("PluginManifest", input, options);
  if (!result.ok) {
    throw new SchemaValidationException(result.error);
  }
  return { migrated: false, value: input };
}

function validateNode(
  node: SchemaNode,
  input: unknown,
  path: string,
  root: SchemaNode,
): ValidationResult<unknown> {
  if (typeof node.$ref === "string") {
    const reference = resolveReference(root, node.$ref);
    return reference
      ? validateNode(reference, input, path, root)
      : failure("schema-reference-unavailable", path);
  }

  if ("const" in node && input !== node.const) {
    return failure("schema-const-mismatch", path);
  }
  if (Array.isArray(node.enum) && !node.enum.includes(input)) {
    return failure("schema-enum-mismatch", path);
  }

  switch (node.type) {
    case "object":
      return validateObject(node, input, path, root);
    case "array":
      return validateArray(node, input, path, root);
    case "string":
      return validateString(node, input, path);
    case "integer":
      return validateNumber(node, input, path, true);
    case "number":
      return validateNumber(node, input, path, false);
    case "boolean":
      return typeof input === "boolean"
        ? success(input)
        : failure("schema-type-mismatch", path);
    default:
      return failure("schema-type-unsupported", path);
  }
}

function validateObject(
  node: SchemaNode,
  input: unknown,
  path: string,
  root: SchemaNode,
): ValidationResult<unknown> {
  if (!isRecord(input)) {
    return failure("schema-type-mismatch", path);
  }

  const properties = asRecord(node.properties) ?? {};
  const required = Array.isArray(node.required) ? node.required : [];
  for (const field of required) {
    if (typeof field === "string" && !(field in input)) {
      return failure("schema-required-field", childPath(path, field));
    }
  }

  for (const [field, value] of Object.entries(input)) {
    const propertySchema = properties[field];
    if (isSchemaNode(propertySchema)) {
      const result = validateNode(
        propertySchema,
        value,
        childPath(path, field),
        root,
      );
      if (!result.ok) {
        return result;
      }
      continue;
    }

    if (node.additionalProperties === false) {
      return failure("schema-unknown-field", childPath(path, field));
    }
    if (isSchemaNode(node.additionalProperties)) {
      const result = validateNode(
        node.additionalProperties,
        value,
        childPath(path, field),
        root,
      );
      if (!result.ok) {
        return result;
      }
    }
  }

  const propertyNames = asRecord(node.propertyNames);
  if (typeof propertyNames?.pattern === "string") {
    for (const field of Object.keys(input)) {
      if (!matchesSupportedPattern(field, propertyNames.pattern)) {
        return failure(
          "schema-invalid-extension-namespace",
          childPath(path, field),
        );
      }
    }
  }

  return success(input);
}

function validateArray(
  node: SchemaNode,
  input: unknown,
  path: string,
  root: SchemaNode,
): ValidationResult<unknown> {
  if (!Array.isArray(input)) {
    return failure("schema-type-mismatch", path);
  }
  if (typeof node.minItems === "number" && input.length < node.minItems) {
    return failure("schema-min-items", path);
  }
  if (!isSchemaNode(node.items)) {
    return failure("schema-type-unsupported", path);
  }

  for (const [index, item] of input.entries()) {
    const result = validateNode(node.items, item, `${path}[${index}]`, root);
    if (!result.ok) {
      return result;
    }
  }
  return success(input);
}

function validateString(
  node: SchemaNode,
  input: unknown,
  path: string,
): ValidationResult<unknown> {
  if (typeof input !== "string") {
    return failure("schema-type-mismatch", path);
  }
  if (
    typeof node.minLength === "number" &&
    [...input].length < node.minLength
  ) {
    return failure("schema-min-length", path);
  }
  return success(input);
}

function validateNumber(
  node: SchemaNode,
  input: unknown,
  path: string,
  integer: boolean,
): ValidationResult<unknown> {
  if (
    typeof input !== "number" ||
    !Number.isFinite(input) ||
    (integer && !Number.isInteger(input))
  ) {
    return failure("schema-type-mismatch", path);
  }
  if (typeof node.minimum === "number" && input < node.minimum) {
    return failure("schema-minimum", path);
  }
  if (typeof node.maximum === "number" && input > node.maximum) {
    return failure("schema-maximum", path);
  }
  return success(input);
}

function validateRequiredExtensions(
  input: unknown,
  knownExtensions: readonly string[],
): ValidationResult<unknown> {
  if (!isRecord(input) || !isRecord(input.compatibility)) {
    return failure("schema-type-mismatch", "$");
  }
  const requiredExtensions = input.compatibility.requiredExtensions;
  if (requiredExtensions === undefined) {
    return success(input);
  }
  if (!Array.isArray(requiredExtensions)) {
    return failure(
      "schema-type-mismatch",
      "$.compatibility.requiredExtensions",
    );
  }

  const extensions = isRecord(input.extensions) ? input.extensions : {};
  for (const [index, namespace] of requiredExtensions.entries()) {
    if (
      typeof namespace !== "string" ||
      !(namespace in extensions) ||
      !knownExtensions.includes(namespace)
    ) {
      return failure(
        "required-extension-unavailable",
        `$.compatibility.requiredExtensions[${index}]`,
      );
    }
  }
  return success(input);
}

function resolveReference(
  root: SchemaNode,
  reference: string,
): SchemaNode | undefined {
  if (!reference.startsWith("#/$defs/")) {
    return undefined;
  }
  const definition = reference.slice("#/$defs/".length);
  const definitions = asRecord(root.$defs);
  const candidate = definitions?.[definition];
  return isSchemaNode(candidate) ? candidate : undefined;
}

function matchesSupportedPattern(value: string, pattern: string): boolean {
  switch (pattern) {
    case "^[a-z0-9]+(?:[.-][a-z0-9]+)+$":
      return /^[a-z0-9]+(?:[.-][a-z0-9]+)+$/u.test(value);
    default:
      return false;
  }
}

function cloneJson(input: JsonRecord): JsonRecord {
  return JSON.parse(JSON.stringify(input)) as JsonRecord;
}

function asRecord(input: unknown): JsonRecord | undefined {
  return isRecord(input) ? input : undefined;
}

function isRecord(input: unknown): input is JsonRecord {
  return typeof input === "object" && input !== null && !Array.isArray(input);
}

function isSchemaNode(input: unknown): input is SchemaNode {
  return isRecord(input);
}

function childPath(path: string, field: string): string {
  return `${path}.${field}`;
}

function success<T>(value: T): ValidationResult<T> {
  return { ok: true, value };
}

function failure(code: string, path: string): ValidationResult<never> {
  return { ok: false, error: { code, path } };
}

File: crates/studio-protocol/src/validation.rs
use std::{fmt, sync::OnceLock};

use serde::de::DeserializeOwned;
use serde_json::Value;

use crate::generated::{PluginManifest, SharedContractDocument};

const SCHEMA_SOURCE: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../schemas/studio.shared.v1alpha1.schema.json"
));

static SHARED_SCHEMA: OnceLock<Value> = OnceLock::new();

/// Stable, machine-readable reason returned by the shared schema validator.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ValidationError {
    pub code: &'static str,
    pub path: String,
}

impl ValidationError {
    fn at(code: &'static str, path: impl Into<String>) -> Self {
        Self {
            code,
            path: path.into(),
        }
    }
}

impl fmt::Display for ValidationError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{} at {}", self.code, self.path)
    }
}

impl std::error::Error for ValidationError {}

/// The result of applying a schema-defined, forward-only manifest migration.
#[derive(Debug, Clone, PartialEq)]
pub struct MigrationResult {
    pub migrated: bool,
    pub value: Value,
}

/// Validates a value against one named definition from the canonical schema source.
pub fn validate_definition(
    definition: &str,
    value: &Value,
    known_extensions: &[&str],
) -> Result<(), ValidationError> {
    let schema = shared_schema();
    let definition_schema = schema
        .pointer(&format!("/$defs/{definition}"))
        .ok_or_else(|| ValidationError::at("schema-definition-unavailable", "$"))?;

    validate_node(schema, definition_schema, value, "$")?;

    if definition == "PluginManifest" {
        validate_required_extensions(value, known_extensions)?;
    }

    Ok(())
}

/// Validates and deserializes a value into one of the generated Rust bindings.
pub fn parse_with_definition<T>(
    definition: &str,
    value: Value,
    known_extensions: &[&str],
) -> Result<T, ValidationError>
where
    T: DeserializeOwned,
{
    validate_definition(definition, &value, known_extensions)?;
    serde_json::from_value(value)
        .map_err(|_| ValidationError::at("schema-deserialization-failed", "$"))
}

/// Validates and deserializes a generated plugin manifest binding.
pub fn parse_plugin_manifest(
    value: Value,
    known_extensions: &[&str],
) -> Result<PluginManifest, ValidationError> {
    parse_with_definition("PluginManifest", value, known_extensions)
}

/// Validates and deserializes the fixture document spanning all shared contract types.
pub fn parse_shared_contract_document(
    value: Value,
    known_extensions: &[&str],
) -> Result<SharedContractDocument, ValidationError> {
    parse_with_definition("SharedContractDocument", value, known_extensions)
}

/// Applies the migration declared by the canonical source, then validates the resulting manifest.
pub fn migrate_plugin_manifest(
    value: Value,
    known_extensions: &[&str],
) -> Result<MigrationResult, ValidationError> {
    let mut migrated_value = value;
    let schema_version = migrated_value
        .get("schemaVersion")
        .and_then(Value::as_str)
        .ok_or_else(|| ValidationError::at("schema-required-field", "$.schemaVersion"))?
        .to_owned();

    let migrations = shared_schema()
        .get("x-openab-migrations")
        .and_then(Value::as_array)
        .expect("canonical schema must contain x-openab-migrations");

    let migration = migrations.iter().find(|candidate| {
        candidate.get("definition").and_then(Value::as_str) == Some("PluginManifest")
            && candidate.get("fromSchemaVersion").and_then(Value::as_str) == Some(&schema_version)
    });

    if let Some(migration) = migration {
        let manifest = migrated_value
            .as_object_mut()
            .ok_or_else(|| ValidationError::at("schema-type-mismatch", "$"))?;
        let target = migration
            .get("toSchemaVersion")
            .and_then(Value::as_str)
            .expect("canonical migration must name a target schema version");

        for (from, to) in migration
            .get("rename")
            .and_then(Value::as_object)
            .expect("canonical migration must declare rename instructions")
        {
            let target_field = to
                .as_str()
                .expect("canonical migration rename target must be a string");
            if let Some(value) = manifest.remove(from) {
                if manifest.contains_key(target_field) {
                    return Err(ValidationError::at("schema-migration-conflict", "$"));
                }
                manifest.insert(target_field.to_owned(), value);
            }
        }
        manifest.insert("schemaVersion".to_owned(), Value::String(target.to_owned()));

        validate_definition("PluginManifest", &migrated_value, known_extensions)?;
        return Ok(MigrationResult {
            migrated: true,
            value: migrated_value,
        });
    }

    let current_version = shared_schema()
        .get("x-openab-schema-version")
        .and_then(Value::as_str)
        .expect("canonical schema must declare x-openab-schema-version");
    if schema_version != current_version {
        return Err(ValidationError::at(
            "schema-migration-unavailable",
            "$.schemaVersion",
        ));
    }

    validate_definition("PluginManifest", &migrated_value, known_extensions)?;
    Ok(MigrationResult {
        migrated: false,
        value: migrated_value,
    })
}

fn shared_schema() -> &'static Value {
    SHARED_SCHEMA.get_or_init(|| {
        serde_json::from_str(SCHEMA_SOURCE)
            .expect("canonical shared schema source must be valid JSON")
    })
}

fn validate_node(
    root: &Value,
    node: &Value,
    value: &Value,
    path: &str,
) -> Result<(), ValidationError> {
    if let Some(reference) = node.get("$ref").and_then(Value::as_str) {
        let referenced = root
            .pointer(reference.trim_start_matches('#'))
            .ok_or_else(|| ValidationError::at("schema-reference-unavailable", path))?;
        return validate_node(root, referenced, value, path);
    }

    if let Some(constant) = node.get("const") {
        if value != constant {
            return Err(ValidationError::at("schema-const-mismatch", path));
        }
    }

    if let Some(allowed) = node.get("enum").and_then(Value::as_array) {
        if !allowed.contains(value) {
            return Err(ValidationError::at("schema-enum-mismatch", path));
        }
    }

    match node.get("type").and_then(Value::as_str) {
        Some("object") => validate_object(root, node, value, path),
        Some("array") => validate_array(root, node, value, path),
        Some("string") => validate_string(node, value, path),
        Some("integer") => validate_integer(node, value, path),
        Some("number") => validate_number(node, value, path),
        Some("boolean") if value.is_boolean() => Ok(()),
        Some("boolean") => Err(ValidationError::at("schema-type-mismatch", path)),
        Some(_) | None => Err(ValidationError::at("schema-type-unsupported", path)),
    }
}

fn validate_object(
    root: &Value,
    node: &Value,
    value: &Value,
    path: &str,
) -> Result<(), ValidationError> {
    let object = value
        .as_object()
        .ok_or_else(|| ValidationError::at("schema-type-mismatch", path))?;
    let properties = node
        .get("properties")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();

    for required in node
        .get("required")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        let property = required
            .as_str()
            .expect("canonical required entry must be a string");
        if !object.contains_key(property) {
            return Err(ValidationError::at(
                "schema-required-field",
                child_path(path, property),
            ));
        }
    }

    for (property, child_value) in object {
        if let Some(property_schema) = properties.get(property) {
            validate_node(
                root,
                property_schema,
                child_value,
                &child_path(path, property),
            )?;
            continue;
        }

        match node.get("additionalProperties") {
            Some(Value::Bool(false)) => {
                return Err(ValidationError::at(
                    "schema-unknown-field",
                    child_path(path, property),
                ));
            }
            Some(Value::Bool(true)) | None => {}
            Some(schema) => validate_node(root, schema, child_value, &child_path(path, property))?,
        }
    }

    if let Some(pattern) = node
        .get("propertyNames")
        .and_then(|property_names| property_names.get("pattern"))
        .and_then(Value::as_str)
    {
        for property in object.keys() {
            if !matches_supported_pattern(property, pattern) {
                return Err(ValidationError::at(
                    "schema-invalid-extension-namespace",
                    child_path(path, property),
                ));
            }
        }
    }

    Ok(())
}

fn validate_array(
    root: &Value,
    node: &Value,
    value: &Value,
    path: &str,
) -> Result<(), ValidationError> {
    let items = value
        .as_array()
        .ok_or_else(|| ValidationError::at("schema-type-mismatch", path))?;
    if let Some(minimum) = node.get("minItems").and_then(Value::as_u64) {
        if items.len() < minimum as usize {
            return Err(ValidationError::at("schema-min-items", path));
        }
    }

    let item_schema = node
        .get("items")
        .expect("canonical array schema must declare items");
    for (index, item) in items.iter().enumerate() {
        validate_node(root, item_schema, item, &format!("{path}[{index}]"))?;
    }
    Ok(())
}

fn validate_string(node: &Value, value: &Value, path: &str) -> Result<(), ValidationError> {
    let string = value
        .as_str()
        .ok_or_else(|| ValidationError::at("schema-type-mismatch", path))?;
    if let Some(minimum) = node.get("minLength").and_then(Value::as_u64) {
        if string.chars().count() < minimum as usize {
            return Err(ValidationError::at("schema-min-length", path));
        }
    }
    Ok(())
}

fn validate_integer(node: &Value, value: &Value, path: &str) -> Result<(), ValidationError> {
    let integer = value
        .as_i64()
        .ok_or_else(|| ValidationError::at("schema-type-mismatch", path))?;
    if let Some(minimum) = node.get("minimum").and_then(Value::as_i64) {
        if integer < minimum {
            return Err(ValidationError::at("schema-minimum", path));
        }
    }
    if let Some(maximum) = node.get("maximum").and_then(Value::as_i64) {
        if integer > maximum {
            return Err(ValidationError::at("schema-maximum", path));
        }
    }
    Ok(())
}

fn validate_number(node: &Value, value: &Value, path: &str) -> Result<(), ValidationError> {
    let number = value
        .as_f64()
        .ok_or_else(|| ValidationError::at("schema-type-mismatch", path))?;
    if let Some(minimum) = node.get("minimum").and_then(Value::as_f64) {
        if number < minimum {
            return Err(ValidationError::at("schema-minimum", path));
        }
    }
    if let Some(maximum) = node.get("maximum").and_then(Value::as_f64) {
        if number > maximum {
            return Err(ValidationError::at("schema-maximum", path));
        }
    }
    Ok(())
}

fn validate_required_extensions(
    value: &Value,
    known_extensions: &[&str],
) -> Result<(), ValidationError> {
    let required_extensions = value
        .pointer("/compatibility/requiredExtensions")
        .and_then(Value::as_array)
        .into_iter()
        .flatten();
    let declared_extensions = value.get("extensions").and_then(Value::as_object);

    for (index, required_extension) in required_extensions.enumerate() {
        let namespace = required_extension
            .as_str()
            .expect("canonical required extension must be a string");
        let is_declared =
            declared_extensions.is_some_and(|extensions| extensions.contains_key(namespace));
        let is_known = known_extensions.contains(&namespace);
        if !is_declared || !is_known {
            return Err(ValidationError::at(
                "required-extension-unavailable",
                format!("$.compatibility.requiredExtensions[{index}]"),
            ));
        }
    }
    Ok(())
}

fn matches_supported_pattern(value: &str, pattern: &str) -> bool {
    match pattern {
        "^[a-z0-9]+(?:[.-][a-z0-9]+)+$" => {
            let mut segment_has_character = false;
            let mut separator_seen = false;
            for character in value.chars() {
                if character.is_ascii_lowercase() || character.is_ascii_digit() {
                    segment_has_character = true;
                } else if (character == '.' || character == '-') && segment_has_character {
                    separator_seen = true;
                    segment_has_character = false;
                } else {
                    return false;
                }
            }
            separator_seen && segment_has_character
        }
        _ => false,
    }
}

fn child_path(path: &str, property: &str) -> String {
    format!("{path}.{property}")
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{matches_supported_pattern, validate_definition};

    #[test]
    fn extension_namespace_pattern_is_strict() {
        assert!(matches_supported_pattern(
            "io.example.telemetry",
            "^[a-z0-9]+(?:[.-][a-z0-9]+)+$"
        ));
        assert!(!matches_supported_pattern(
            "io.example..telemetry",
            "^[a-z0-9]+(?:[.-][a-z0-9]+)+$"
        ));
    }

    #[test]
    fn strict_base_schema_rejects_unknown_fields() {
        let result = validate_definition(
            "FleetIdentity",
            &json!({
                "id": "flt_1",
                "displayName": "Fleet",
                "revision": "rev_1",
                "untrusted": true
            }),
            &[],
        );
        assert_eq!(
            result.expect_err("must reject unknown base field").code,
            "schema-unknown-field"
        );
    }
}
